lib/allocation.c

```diff
@@ -1113,20 +1118,29 @@ hw_alloc_assign(const unsigned technology,
                                 goto pqos_alloc_assign_exit;
                         }
                         l2id = pi->l2_id;
-                } else {
-                        if (i != 0 && socket != pi->socket) {
+		} else if (mba_req) {
+			if (i != 0 && mba_id != pi->mba_id) {
+				ret = PQOS_RETVAL_PARAM;
+				goto pqos_alloc_assign_exit;
+			}
+			mba_id = pi->mba_id;
+		} else {
+			if (i != 0 && l3cat_id != pi->l3cat_id) {
                                 ret = PQOS_RETVAL_PARAM;
                                 goto pqos_alloc_assign_exit;
                         }
-                        socket = pi->socket;
+			l3cat_id = pi->l3cat_id;
                 }
         }
 
         /* find an unused class from highest down */
-        if (!l2_req)
-                ret = get_unused_cos(socket, technology, class_id);
-        else
-                ret = get_unused_cos(l2id, technology, class_id);
+	if (l2_req)
```

Comment:
For each technology we might have different number of classes available. When 2 technologies are requested we need to obtain class number that supports both technologies.

Single call to get_unused_cos should do the trick:
```suggestion
	ret = get_unused_cos(resource_id, technology, class_id);
```

Reply:
ok. done